wrong location on import error message reference to undefined rule

Error message should point to location where rule was referenced in imported file, in this case Nested.g4

[mpawlowsky@vmott-mpawlowsky-seas error]$ ant
Buildfile: /home/mpawlowsky/error/build.xml

generate:
     [java] error(56): error.g4:4:8: reference to undefined rule: c

BUILD FAILED
/home/mpawlowsky/error/build.xml:3: Java returned: 1

Total time: 0 seconds
[mpawlowsky@vmott-mpawlowsky-seas error]$ cat error.g4 
grammar Error;
import Nested;

[mpawlowsky@vmott-mpawlowsky-seas error]$ cat Nested.g4 
grammar Nested;

a:
  'a' | c;

 <project name="error" default="generate">
  <target name="generate" >
                <java classname="org.antlr.v4.Tool" fork="false" failonerror="true">
                        <arg value="-o" />
                        <arg value="java" />
                        <arg value="-package" />
                        <arg value="com.example" />
                        <arg value="-visitor" />
                        <arg value="-Werror" />
                        <arg value="error.g4" />
                        <classpath>
                                <pathelement location="antlr-4.4-complete.jar"/>
                        </classpath>
                </java>
</target>
</project>

--- a/tool/src/org/antlr/v4/Tool.java
+++ b/tool/src/org/antlr/v4/Tool.java
@@ -480,18 +480,18 @@ public class Tool {
 			@Override
 			public void ruleRef(GrammarAST ref, ActionAST arg) {
 				RuleAST ruleAST = ruleToAST.get(ref.getText());
+				String fileName = ref.getToken().getInputStream().getSourceName();
 				if (Character.isUpperCase(currentRuleName.charAt(0)) &&
 					Character.isLowerCase(ref.getText().charAt(0)))
 				{
 					badref = true;
-					String fileName = ref.getToken().getInputStream().getSourceName();
 					errMgr.grammarError(ErrorType.PARSER_RULE_REF_IN_LEXER_RULE,
 										fileName, ref.getToken(), ref.getText(), currentRuleName);
 				}
 				else if ( ruleAST==null ) {
 					badref = true;
 					errMgr.grammarError(ErrorType.UNDEFINED_RULE_REF,
-										g.fileName, ref.token, ref.getText());
+										fileName, ref.token, ref.getText());
 				}
 			}
 			@Override

--- a/tool/test/org/antlr/v4/test/tool/TestCompositeGrammars.java
+++ b/tool/test/org/antlr/v4/test/tool/TestCompositeGrammars.java
@@ -30,10 +30,13 @@
 
 package org.antlr.v4.test.tool;
 
+import org.antlr.v4.tool.ANTLRMessage;
 import org.antlr.v4.tool.ErrorType;
 import org.antlr.v4.tool.Grammar;
 import org.antlr.v4.tool.GrammarSemanticsMessage;
 import org.junit.Test;
+
+import java.io.File;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
@@ -58,6 +61,25 @@ public class TestCompositeGrammars extends BaseTest {
 		writeFile(tmpdir, "M.g4", master);
 		ErrorQueue equeue = antlr("M.g4", false, "-lib", subdir);
 		assertEquals(equeue.size(), 0);
+	}
+
+	@Test public void testErrorInImportedGetsRightFilename() throws Exception {
+		String slave =
+			"parser grammar S;\n" +
+			"a : 'a' | c;\n";
+		mkdir(tmpdir);
+		writeFile(tmpdir, "S.g4", slave);
+		String master =
+			"grammar M;\n" +
+			"import S;\n";
+		writeFile(tmpdir, "M.g4", master);
+		ErrorQueue equeue = antlr("M.g4", false, "-lib", tmpdir);
+		ANTLRMessage msg = equeue.errors.get(0);
+		assertEquals(ErrorType.UNDEFINED_RULE_REF, msg.getErrorType());
+		assertEquals("c", msg.getArgs()[0]);
+		assertEquals(2, msg.line);
+		assertEquals(10, msg.charPosition);
+		assertEquals("S.g4", new File(msg.fileName).getName());
 	}
 
 	@Test public void testImportFileNotSearchedForInOutputDir() throws Exception {

